### ISSUE TITLE
allow_reuse=True bug

### ISSUE THREAD

**Issue by snelso1** [Azure/MachineLearningNotebooks#379]
We have seen a bug lately whereby if we set allow_reuse=True on all of our pipeline steps (see below for an example), then our pipeline starts failing with errors that are not very descriptive (see attached files).  This issue started happening for us on Monday when a colleague updated a CSV in the blob which the step depended on.  We can get the pipeline working again by setting allow_reuse=False, but if we set it back to True then it starts failing again.  Let me know if there is more information I can provide to troubleshoot.

munge_time_step=PythonScriptStep(
    name='munge time',
    script_name='munge_time.py',
    arguments=['--input_dir', input_dir,
               '--munged_hc_dir', munged_hc,
               '--output_dir', munged_time],
    compute_target=compute_target,
    inputs=[input_dir, munged_hc],
    outputs=[munged_time],
    runconfig=aml_compute,
    source_directory=os.path.join(os.getcwd(), 'compute'),
    allow_reuse=pipeline_reuse
)

[55_batchai_exec.txt](https://github.com/Azure/MachineLearningNotebooks/files/3184582/55_batchai_exec.txt)
[executionlog.txt](https://github.com/Azure/MachineLearningNotebooks/files/3184583/executionlog.txt)

**Comment by swanderz** [comment]
cc: @heatherbshapiro @sanpil @jpe316 
This is the same bug we were seeing from that email thread from two weeks ago.
This time we're all on `1.0.10` so that theory's out the window now.
New theory: the run that is `allow_reuse` is trying to re-hydrate is corrupted or something...
We've opened a priority support ticket for it. Pretty big blocker for us right now

**Comment by yanrez** [comment]
Most likely when you set it to allow_reuse it picks up old data which had some problem throwing off your script and making it fail. Did you look into content of the input - do you see data having some bug?

**Comment by snelso1** [comment]
We know that the content in the source location is good, it's just that whenever the step re-runs due to the content changing fails before it can even get to our python code (no driver_log).  Any other idea we can try to troubleshoot?

**Comment by yuezengms** [comment]
It seems job exited with non-zero code. Could you please attach driver log which may have detailed information?
2019-05-14T21:21:03Z launching Custom job
2019-05-14T21:21:11Z job exited with code 1

**Comment by snelso1** [comment]
Hi, no driver log gets generated, please see attached image that shows the logs that I see.  All pertinent info was included in the previous log files which doesn't really include any helpful troubleshooting info.

Thanks,

![azureml_reuse_bug](https://user-images.githubusercontent.com/1169037/57890249-f6042f00-77eb-11e9-9b48-e89a9a4462a2.png)

Steve

**Comment by gargiulman** [comment]
hello,

Could you please help clarify on below  --

1) When allow_reuse = True and data in the input file to the step changes  from say X rows to Y rows. 
Here i would ideally expect the step to re-run whereas it doesnt and picks data from prev run.

2) When allow_reuse = True and i make change in the script file of one step. where my pipeline has many steps. Now i see that all the steps start re-running whereas ideally only the step whose script changed should run. This was my understanding. Also there is no dependency between the steps, which can force them to re-run.

**Comment by sanpil** [comment]
1. See https://github.com/Azure/MachineLearningNotebooks/issues/298 (If the data is in a datastore, we would not be able to detect the data change. If the data is uploaded as part of the snapshot (under source_directory) [this is not recommended though], then the hash will change and will trigger a rerun.)

2. Ideally, you should use separate source_directory for separate steps. If all the scripts in your steps are in a single directory, the hash of that directory changes every time you make a change to one script forcing all steps to rerun. If you look at [https://aka.ms/pl-get-started](https://aka.ms/pl-get-started), you can see that 3 separate sub-directories are used to keep the three scripts used in the three steps.

**Comment by gargiulman** [comment]
Thanks for the response. For point 2 , below is my scenario -

a) I created a new branch from master repo which has allow_reuse= true 
b) then i made changes in my branch, tested it and ran the experiment. I see for the 1st run, all steps got created and that was expected, because my script file changed and as per your comment, the hash of the directory changed. 
 
I ran the experiment again, and this time results from prev run were used. This was also expected. 

c) Now i went raising a Pull Request to merge to master.  I had set pre-merging validation, so my build pipeline for UT started running which actually runs my experiment. I was expecting this time my that results from previous run should be used. But i see the steps again getting created. 

Any guidance why this is happening. The scripts havent changed since i tested on local and then raising a PR. Allow_resuse = true is also set.
